cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -695,6 +702,65 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
     }
   }
 
+  /**
+   * Delete the business metadata for the application and the programs.
+   */
+  private void deleteAppBusinessMetadata(Id.Application appId) {
+    // Remove metadata for the Application itself.
+    businessMds.removeProperties(appId);
+    businessMds.removeTags(appId);
+
+    // Remove business metadata for the programs of the Application
+    // TODO: Need to remove this we support prefix search of metadata type.
+    ApplicationSpecification appSpec = store.getApplication(appId);
+    for (ProgramType programType : ProgramType.values()) {
+      deleteProgramBusinessMetadataForApp(programType, appId, appSpec);
+    }
+  }
+
+  /**
+   *  Delete the business metadata for a program type in an application.
+   *  TODO: Need to remove this we support prefix search of metadata type.
+   */
+  private void deleteProgramBusinessMetadataForApp(ProgramType programType, Id.Application appId,
+                                                   ApplicationSpecification appSpec) {
+    Set<String> programNames = null;
+    switch (programType) {
+      case FLOW:
+        programNames = appSpec.getFlows().keySet();
+        break;
+      case MAPREDUCE:
+        programNames = appSpec.getMapReduce().keySet();
+        break;
+      case WORKFLOW:
+        programNames = appSpec.getWorkflows().keySet();
+        break;
+      case SERVICE:
+        programNames = appSpec.getServices().keySet();
+        break;
+      case SPARK:
+        programNames = appSpec.getSpark().keySet();
+        break;
+      case WORKER:
+        programNames = appSpec.getWorkers().keySet();
+        break;
+      default:
+        LOG.info("Unable to remove business metadata for unsupported program type: " + programType.name());
+        break;
+    }
+
+    if (programNames == null) {
+      LOG.debug("No program metadata removed for application {}", appId.toString());
+      return;
+    }
+
+    for (String programName : programNames) {
```

Comment:
Doesn't this also require notifications of the change? How are the notifications sent? Perhaps all meta data notifications should go through the metadata service?